mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

```diff
@@ -336,32 +444,156 @@ struct ConvertVectorStore final : OpConversionPattern<vector::StoreOp> {
-        loc, VectorType::get(numElements, newElementType),
-        op.getValueToStore());
+    auto foldedNumFrontPadElems =
+        isUnalignedEmulation
+            ? getConstantIntValue(linearizedInfo.intraDataOffset)
+            : 0;
 
-    rewriter.replaceOpWithNewOp<vector::StoreOp>(
-        op, bitCast.getResult(), adaptor.getBase(),
-        getValueOrCreateConstantIndexOp(rewriter, loc, linearizedIndices));
+    if (!foldedNumFrontPadElems) {
+      // Unimplemented case for dynamic front padding size != 0
+      return failure();
+    }
+
+    auto linearizedMemref = cast<MemRefValue>(adaptor.getBase());
+
+    // Shortcut: conditions when subbyte store at the front is not needed:
+    // 1. The source vector size is multiple of byte size
+    // 2. The address of the store is aligned to the emulated width boundary
+    if (!isUnalignedEmulation && *foldedNumFrontPadElems == 0) {
+      auto numElements = origElements / numSrcElemsPerDest;
+      auto bitCast = rewriter.create<vector::BitCastOp>(
+          loc, VectorType::get(numElements, newElementType),
+          op.getValueToStore());
+      rewriter.replaceOpWithNewOp<vector::StoreOp>(
+          op, bitCast.getResult(), linearizedMemref,
+          getValueOrCreateConstantIndexOp(rewriter, loc, linearizedIndices));
+      return success();
+    }
+
+    // The index into the target memref we are storing to
+    Value currentDestIndex =
+        getValueOrCreateConstantIndexOp(rewriter, loc, linearizedIndices);
+    auto constantOne = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    auto subWidthStoreMaskType =
+        VectorType::get({numSrcElemsPerDest}, rewriter.getI1Type());
+    // The index into the source vector we are currently processing
+    auto currentSourceIndex = 0;
+
+    // 1. Partial width store for the first byte, when the store address is not
+    // aligned to emulated width boundary, deal with the unaligned part so that
+    // the rest elements are aligned to width boundary.
+    auto frontSubWidthStoreElem =
+        (numSrcElemsPerDest - *foldedNumFrontPadElems) % numSrcElemsPerDest;
+    if (frontSubWidthStoreElem > 0) {
+      SmallVector<bool> frontMaskValues(numSrcElemsPerDest, false);
+      if (*foldedNumFrontPadElems + origElements < numSrcElemsPerDest) {
+        std::fill_n(frontMaskValues.begin() + *foldedNumFrontPadElems,
+                    origElements, true);
+        frontSubWidthStoreElem = origElements;
+      } else {
+        std::fill_n(frontMaskValues.end() - frontSubWidthStoreElem,
+                    *foldedNumFrontPadElems, true);
+      }
+      auto frontMask = rewriter.create<arith::ConstantOp>(
+          loc, DenseElementsAttr::get(subWidthStoreMaskType, frontMaskValues));
+
+      currentSourceIndex = numSrcElemsPerDest - (*foldedNumFrontPadElems);
+      auto value =
+          extractSliceIntoByte(rewriter, loc, valueToStore, 0,
+                               frontSubWidthStoreElem, *foldedNumFrontPadElems);
```

Comment:
Below is a snippet from the function. Is it only correct for `memref-load-bitwidth=8` case? Should we pass the number of the bits of the memref element type (i.e., `dstBits`) to the function? On most CPUs, it is correct. On SPIR-V, it could be not enough. On some GPU targets, they could only support i32 types, but not i8 types. This is why we make the `memref-load-bitwidth` configurable.  

```
auto scale = 8 / vectorElementType.getIntOrFloatBitWidth();
```

Reply:
If we are passing `dstBits` to the function. I think we can rename it as `extractSliceIntoBytes`, i.e., adding `s` at the end.

Reply:
@hanhanW I think this issue is relevant in other similar functions in the same file, and it has been like this for a while. We should address that in a subsequent patch, and make them consistent.

> On some GPU targets, they could only support i32 types, but not i8 types

`i8` types are emulated as `i32` types on those targets, so memref loads of `i8` will be extended as loads of `i32`. When unaligned loads are not supported, some shiftings and maskings might be inserted to extract the value.

So in this sense, supporting only `biwidth == 8` in this pass as a whole is okay, but with lower performance. But the issue of `memref-load-bit-width != 8` might pose inconsistency issues. We should fix that (in all those functions in the same file and correct namings to resolve confusions as well).

Reply:
Just wanna make sure my understanding is correct. I agree that we can address it in a subsequent patch. However, I do not follow why we have the issue in other functions? The below test file has the tests for `memref-load-bit-width=i32` config.

https://github.com/llvm/llvm-project/blob/b446c208a5f0e2ad7193cc23e70642d207db4d13/mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir#L2

I do see that there are no i32 tests in the [vector-emulate-narrow-type-unaligned.mlir](https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir). Are you saying that i32 is not supported in unaligned cases, and we should fix it in follow-ups? If so, it makes sense to me.